File: 2025/1/week1/안정적인 네트워크_백준_2406/Solution.java
import java.io.*;
import java.util.*;

class Solution{
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static StringTokenizer st;
    static int n,m;
    static UnionFind unionFind;
    static PriorityQueue<int[]> connectionCost;
    static HashSet<Integer> connectedNetworks;
    static int connectedCount = 0;
    static int minimumCost = 0;
    static int pairNum = 0;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException{
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        connectionCost = new PriorityQueue<int[]>((o1, o2) -> {
            if(o1[0]==o2[0])return o1[1] - o2[1];
            return o1[0] - o2[0];
        });
        connectedNetworks = new HashSet<>();

        unionFind = new UnionFind(n);
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken())-1;
            int y = Integer.parseInt(st.nextToken())-1;
            connectedNetworks.add(x*1_000+y);
            if(unionFind.union(x, y))connectedCount++;
        }

        for (int i = 0; i < n; i++){
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < i; j ++){
                int cost = Integer.parseInt(st.nextToken());
                if(j == 0)continue;
                int intHash = i*1_000+j;
                if(connectedNetworks.contains(intHash))continue;
                
                connectionCost.add(new int[]{cost,intHash});
            }
        }

        while (!connectionCost.isEmpty()) {
            int[] node = connectionCost.poll();
            int x = node[1]/1_000;
            int y = node[1]%1_000;
            if(unionFind.union(x, y)){
                minimumCost+=node[0];
                pairNum+=1;
                sb.append(x+1).append(" ").append(y+1).append('\n');
                connectedCount++;
            }
            if(connectedCount>=n-2)break;
        }
        System.out.println(""+minimumCost+" "+pairNum);
        System.out.println(sb);
    }
}

class UnionFind{
    int size = 0;
    int[] parentRank;

    public UnionFind(int size){
        this.size = size;
        parentRank = new int[size];
        Arrays.fill(parentRank, -1);
    }

    public int find(int x){
        if(parentRank[x] < 0) return x;
        return parentRank[x] = find(parentRank[x]);
    }

    public boolean union(int x, int y){
        int xRoot = find(x);
        int yRoot = find(y);
        if(xRoot == yRoot) return false;
        if(parentRank[xRoot] > parentRank[yRoot]){
            parentRank[yRoot] += parentRank[xRoot];
            parentRank[xRoot] = yRoot;
        }else{
            parentRank[xRoot] += parentRank[yRoot];
            parentRank[yRoot] = xRoot;
        }
        return true;
    }
}

File: 2025/1/week1/안정적인 네트워크_백준_2406/Solution.py
import sys
import heapq

class UnionFind:
    def __init__(self, size):
        self.parentRank = [ -1 for _ in range(size)]
        
    def find(self, x):
        if self.parentRank[x] < 0:
            return x
        self.parentRank[x] = self.find(self.parentRank[x])
        return self.parentRank[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX == rootY:
            return False
        
        if self.parentRank[rootX] > self.parentRank[rootY]:
            self.parentRank[rootY]+=self.parentRank[rootX]
            self.parentRank[rootX]=rootY
        else:
            self.parentRank[rootX]+=self.parentRank[rootY]
            self.parentRank[rootY]=rootX
        return True

n,m = map(int,sys.stdin.readline().split())
unionFind = UnionFind(n)
pq = []
connected = set()
connectedCount = 0
minimumCost = 0
pairNum = 0
answer = ""

for i in range(m):
    x,y = map(lambda a: int(a)-1, sys.stdin.readline().split())
    if unionFind.union(x,y):
        connectedCount+=1
    connected.add(1_000*x+y)

for i in range(n):
    costs = list(map(lambda a: int(a), sys.stdin.readline().split()))
    for j in range(i):
        if j == 0:
            continue
        if (i*1_000+j) in connected:
            continue
        heapq.heappush(pq,[costs[j],i*1_000+j])

while connectedCount < n-2 and len(pq)>0:
    cost, index = heapq.heappop(pq)
    # print(cost, index)
    x = index//1_000
    y = index%1_000
    if unionFind.union(x,y):
        connectedCount+=1
        minimumCost+=cost
        pairNum+=1
        answer+=str(x+1)+" "+str(y+1)+"\n"

print(str(minimumCost)+" "+str(pairNum))
print(answer)